Developer docs for caching (#4211)
numba/numba: docs/source/developer/caching.rst

Features Compatible with Caching
--------------------------------

The following features are explicitly verified to work with caching.

- ufuncs and gufuncs for the ``cpu`` and ``parallel`` target
- parallel accelerator features (i.e. ``parallel=True``)

Caching Limitations
-------------------

This is a list of known limitation of the cache:

- Cache invalidation fails to recognize changes in symbols defined in a
  different file.
- Global variables are treated as constants. The cache will remember the value
  in the global variable used at compilation. On cache load, the cached
  function will not rebind to the new value of the global variable.

.. _cache-sharing:

Cache Sharing
-------------

It is safe to share and reuse the contents in the cache directory on a
different machine. The cache remembers the CPU model and the available
CPU features during compilation. If the CPU model and the CPU features do
not match exactly, the cache contents will not be considered.
(Also see :envvar:`NUMBA_CPU_NAME`)

If the cache directory is shared on a network filesystem, concurrent
read/write of the cache is safe only if file replacement operation is atomic

> for the filesystem. Numba always write to a unique temporary file first, it then

Reviewer: Should be `Numba always writes`, but that'll cause a line wrap needed so am not using the suggester thing.

- The LLVM module must be *self-contained*, meaning that it cannot rely on
  other compiled units without linking to them.
- The only allowed external symbols are from the
  :ref:`NRT <arch-numba-runtime>` or other common symbols from system libraries
  (i.e. libc and libm).

Debugging note:

- Look for the usage of ``inttoptr`` in the LLVM IR or
  ``target_context.add_dynamic_add()`` in the lowering code in Python.
  They indicate potential usage of runtime address. Not all uses are
  problematic and some are necessary. Only the conversion of constant integers
  into pointers will affect caching.
- Misuse of dynamic address or dynamic symbols will likely result in a
  segfault.
- Linking order matters because unused symbols are dropped after linking.
  Linking should start from the leaf nodes of the dependency graph.

Features Compatible with Caching
--------------------------------

The following features are explicitly verified to work with caching.

- ufuncs and gufuncs for the ``cpu`` and ``parallel`` target
- parallel accelerator features (i.e. ``parallel=True``)

Caching Limitations
-------------------

This is a list of known limitation of the cache:

> - Functions using ``hash(str)`` will produce unexpected results when loaded

Reviewer: Is this still valid?

Author: no, removed.

  from cache. This also affects dictionary usage (i.e. ``numba.typed.Dict``).
- Cache invalidation fails to recognize changes in symbols defined in a
  different file.
- Global variables are treated as constants. The cache will remember the value
  in the global variable used at compilation. On cache load, the cached
  function will not rebind to the new value of the global variable.

.. _cache-sharing:

Cache Sharing
-------------

> It is safe to share and reuse the contents in the cache directory in a

Reviewer: ```suggestion
It is safe to share and reuse the contents in the cache directory on a
```

different machine. The cache remembers the CPU model and the available
CPU features during compilation. If the CPU model and the CPU features do
not match exactly, the cache contents will not be considered.
(Also see :envvar:`NUMBA_CPU_NAME`)

If the cache directory is shared on a network filesystem, concurrent
read/write of the cache is safe only if file replacement operation is atomic

> for the filesystem. Numba always write to a unique temporary file first. Then,

Reviewer: ```suggestion
for the filesystem. Numba always write to a unique temporary file first, it then
```

> it replaces the target cache file path with the temporary file. Numba is

Reviewer: ```suggestion
replaces the target cache file path with the temporary file. Numba is
```

> tolerant against lost cache file and lost cache entry.

Reviewer: ```suggestion
tolerant against lost cache files and lost cache entries.
```

.. _cache-clearing:

Cache Clearing
--------------

The cache is invalidated when the corresponding source file is modified.

> However, it is necessary somtimes to clear the cache directory manually.

Reviewer: ```suggestion
However, it is necessary sometimes to clear the cache directory manually.
```

.. _developer-caching:

================
Notes on Caching
================

Numba supports caching of compiled functions into the filesystem for future
use of the same functions.

The Implementation
==================

Caching is done by saving the compiled *object code*, the ELF object of the
executable code.  By using the *object code*, cached functions have minimal

> overhead because no compilation is needed. The cached data is saved under

Reviewer: ```suggestion
overhead because no compilation is needed. The cached data is saved under the
```

``__pycache__`` directory. The index of the cache is stored in a ``.nbi``

> file.  One index per function. It lists all the overloaded signatures compiled

Reviewer: ```suggestion
file, with one index per function, and it lists all the overloaded signatures compiled
```

> for the function. The *object code* is stored in files of ``.nbc`` extension.

Reviewer: ```suggestion
for the function. The *object code* is stored in files with an ``.nbc`` extension,
```

> One per overload.

Reviewer: ```suggestion
one file per overload.
```

Requirements for Cacheability
-----------------------------

> Developers should note the requirements for a function to be cached to ensure

Reviewer: ```suggestion
Developers should note the requirements of a function to permit it to be cached to ensure
```

that the features they are working on are compatible with caching.

Requirements for cacheable function:

- The LLVM module must be *self-contained*, meaning that it cannot rely on
  other compiled units without linking to them.
- The only allowed external symbols are from the
  :ref:`NRT <arch-numba-runtime>` or other common symbols from system libraries
  (i.e. libc and libm).

Debugging note:

- Look for the usage of ``inttoptr`` in the LLVM IR or
  ``target_context.add_dynamic_add()`` in the lowering code in Python.

>   They indicate potential usage of runtime address. Not all usage are

Reviewer 2: ```suggestion
  They indicate potential usage of runtime address. Not all usages are
```

>   They indicate potential usage of runtime address. Not all usages are

Reviewer: ```suggestion
  They indicate potential usage of runtime address. Not all uses are
```

  problematic and some are necessary. Only the conversion of constant integers
  into pointers will affect caching.
- Misuse of dynamic address or dynamic symbols will likely result in a
  segfault.
- Linking order matters because unused symbols are dropped after linking.
  Linking should start from the leaf nodes of the dependency graph.

> Features compatible with caching

Reviewer 2: ```suggestion
Features Compatible with Caching
```

To make the subtitle capitalization consistent.

Caching is done by saving the compiled *object code*, the ELF object of the
executable code.  By using the *object code*, cached functions have minimal
overhead because no compilation is needed. The cached data is saved under the
``__pycache__`` directory. The index of the cache is stored in a ``.nbi``

> file, with one index per function, and it lists all the overloaded signatures compiled

Reviewer: this line is too long now, sorry, that was the auto-suggest thing not line wrapping

> for the function. The *object code* is stored in files with an ``.nbc`` extension,

Reviewer: and this one

one file per overload.

Requirements for Cacheability
-----------------------------

> Developers should note the requirements of a function to permit it to be cached to ensure

Reviewer: and this one
